src/linear_algebra/clifford_algebra.lean

```diff
@@ -0,0 +1,155 @@
+/-
+Copyright (c) 2020 Eric Wieser. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Eric Wieser, Utensil Song.
+-/
+
+import algebra.ring_quot
+import linear_algebra.tensor_algebra
+import linear_algebra.quadratic_form
+
+/-!
+# Clifford Algebras
+
+We construct the Clifford algebra of a module `M` over a commutative ring `R`, equipped with
+a quadratic_form `Q`.
+
+## Notation
+
+The Clifford algebra of the `R`-module `M` equipped with a quadratic_form `Q` is denoted as
+`clifford_algebra R M Q`.
+
+Given a linear morphism `f : M → A` from a semimodule `M` to another `R`-algebra `A`, such that
+`cond : ∀ m, f m * f m = algebra_map _ _ (Q m)`, there is a (unique) lift of `f` to an `R`-algebra
+morphism, which is denoted `clifford_algebra.lift R Q f cond`.
+
+The canonical linear map `M → clifford_algebra R M Q` is denoted `clifford_algebra.ι R Q`.
+
+## Theorems
+
+The main theorems proved ensure that `clifford_algebra R M Q` satisfies the universal property
+of the clifford algebra.
+1. `ι_comp_lift` is the fact that the composition of `ι R Q` with `lift R Q f cond` agrees with `f`.
+2. `lift_unique` ensures the uniqueness of `lift R Q f cond` with respect to 1.
+
+## Implementation details
+
+The clifford algebra of `M` is constructed as a quotient of the tensor algebra, as follows.
+1. We define a relation `clifford_algebra.rel R M Q` on `tensor_algebra R M`.
+   This is the smallest relation which identifies squares of elements of `M` with `Q m`.
+2. The clifford algebra is the quotient of the tensor algebra by this relation.
+
+This file is almost identical to `linear_algebra/exterior_algebra.lean`.
+Indeed, the exterior algebra can be recovered by setting `Q` to `0`.
+-/
+
+variables (R : Type*) [comm_ring R]
+variables (M : Type*) [add_comm_group M] [module R M]
```

Comment:
I think `R` and `M` can probably be made implicit in this file, since lean can deduce what they are from `Q`. I think it would be much nicer to write `clifford_algebra Q` as opposed to `clifford_algebra R M Q`.

Reply:
I hadn't thought of that, good idea

Reply:
I've been struggling what Lean can deduce, so I expect problems once they are declared implicit. Have you try to do so? @eric-wieser 

It seems to me that there's a pattern in mathlib that in the basic sections these variables would be declared as explicit and building blocks are laid out, then in a later namespace/section, when convenient and appropriate,  would be declared implicit and there would be a few enduser-friendly theorems/lemmas.

Reply:
In this case it's clear that `R` and `M` can be made implicit, since they're both part of the type declaration for `Q`. So as long as `Q` is explicit, lean will figure out what the `R` and `M` must be from the type of `Q`. 

A similar thing happens for sets, e.g. once you know about `S : set X` explicitly, the `X` can be made implicit since it's part of the type of `S`.